src/compiler/checker.ts

```diff
@@ -8309,6 +8320,18 @@ namespace ts {
             }
         }
 
+        function isEmptyResolvedType(t: ResolvedType) {
+            return t.properties.length === 0 &&
+                t.callSignatures.length === 0 &&
+                t.constructSignatures.length === 0 &&
+                !t.stringIndexInfo &&
+                !t.numberIndexInfo;
+        }
+
+        function isEmptyObjectType(type: Type) {

```

Comment:
why can't we use (type===emptyObjectType) here? Do we really have to resolve every type and then check its structure?